staging/src/k8s.io/apiserver/pkg/server/openapi/openapi_handler.go

```diff
@@ -0,0 +1,160 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openapi
+
+import (
+	"bytes"
+	"compress/gzip"
+	"encoding/json"
+	"fmt"
+	"mime"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/go-openapi/spec"
+	"github.com/golang/protobuf/proto"
+	"github.com/googleapis/gnostic/OpenAPIv2"
+	"github.com/googleapis/gnostic/compiler"
+	"gopkg.in/yaml.v2"
+	genericmux "k8s.io/apiserver/pkg/server/mux"
+)
+
+type OpenAPIService struct {
+	orgSpec      *spec.Swagger
+	specBytes    []byte
+	specPb       []byte
+	specPbGz     []byte
+	lastModified time.Time
+	updateHooks  []func(*http.Request)
+}
+
+// RegisterOpenAPIService registers a handler to provides standard OpenAPI specification.
+func RegisterOpenAPIService(openapiSpec *spec.Swagger, servePath string, mux *genericmux.PathRecorderMux) (*OpenAPIService, error) {
+
+	if !strings.HasSuffix(servePath, JSON_EXT) {
+		return nil, fmt.Errorf("Serving path must ends with \"%s\".", JSON_EXT)
```

Comment:
The caller to RegisterOpenAPIService will call it with json extension, but the function will reuse the name and serve other formats too. The caller do not need to worry about other formats, only json.